### DISCUSSION TITLE
Replace ZMQ(router/dealer) to NNG in raw mode problem

### DISCUSSION THREAD

**Discussion by mohit-kumawatt** [nanomsg/nng#1791] [Q&A] [answered]
i want to replace zmq by nng in my project and in which there are more than one number of dealers where i want to send the message. Can it be possible to replace zmq. and if it is possible then please provide the document of nng raw mode.

**Comment by gdamore** [comment, marked as answer]
You probably don't need to use raw mode.

You can use nng_device() to build these kinds of configurations.  You only need raw mode if you want to have more direct control over the routing.

The details of the raw mode are specified ind the documentation.  Please see for example the  documentation for REQ in nng_req.7

**Comment by mohit-kumawatt** [comment]
i want to work on nng(router/dealer) . can i done by nng_device.

**Comment by gdamore** [comment]
The "nng_device()" function creates a router/dealer type configuration.  Please see the documentation for it.  The code is fairly straight-forward as well -- although be advised that it is just a very simple proxy, as it does not know how (for example) to select different peers -- the individual sockets that are connected will each still use their normal modes of distribution.  (E.g. round robin for REQ/REP, broadcast for PUB, etc.)

**Comment by lindarean** [comment]
int ZMQGateway::zmqReceiverThread() {
    zmq_pollitem_t zmqItems[] = {
        {zmqRequestSocket, 0, ZMQ_POLLIN, 0},
        {zmqResponseSocket, 0, ZMQ_POLLIN, 0},
    };

    while (run) {
        zmq_poll(zmqItem, 2, -1);
        if (zmqItems[0].revents & ZMQ_POLLIN) {
  ...........................
  this is the first code using ZMQ_POLL
  
  now this is the second code 
  

  int NNGGateway::nngReceiverThread()
{
    // (created using nng_req0_open_raw())
    // nng_socket sockrequest, sockreply;
    while (runNng)
    {
        int rv;
        if ((rv = nng_req0_open_raw(&sockrequest_raw)) != 0)
        {
            std::cerr << "Error opening REQ socket: " << nng_strerror(rv) << std::endl;
            return rv;
        }

        if ((rv = nng_rep0_open_raw(&sockreply_raw)) != 0)
        {
            std::cerr << "Error opening REP socket: " << nng_strerror(rv) << std::endl;
            nng_close(sockrequest_raw);
            return rv;
        }

        // Use nng_device for message forwarding
        int deviceResult = nng_device(sockrequest_raw, sockreply_raw);

        if (deviceResult != 0)
        {
            std::cerr << "Error in nng_device: " << nng_strerror(deviceResult) << std::endl;
            return deviceResult;
        }
    }
    return 0;
}
     i want that nng should work like ZMQ_POLL is it fine without functionality change in the code

**Reply by gdamore** [reply]
I believe the examples you give here achieve that.

**Comment by dabelknap** [comment]
With the ZMQ ROUTER/DEALER it lets you manually specify the identity of the connection. Is it possible to do that with raw NNG sockets? From playing with this example (https://www.darenliang.com/posts/dealer-router-pattern-in-nng/), it appears the identity is still be populated automatically in the header.
